Add rate of change to statistics sensor

flo-wer opened the pull request:
## Description:
This adds rate of change calculations to the statistics sensor.
The following changes were made:
* calculate and export rate of change attribute, i.e. the average change of the sensor value over time.
* add test case for new rate of change attribute.
* max_age/min_age is now always exported as the restriction makes no sense.
* downgrade errors of missing data. They shouldn't be treated as errors as the case of only one or even zero data points can occur intentional, e.g. when a short max_age is used.
* simplify and reorder calculations and attribute assignments.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation:** https://github.com/home-assistant/home-assistant.io/pull/6034

## Example entry for `configuration.yaml` (if applicable):
```yaml
- platform: statistics
  name: water_level_statistics
  entity_id: sensor.water_level
  sampling_size: 3
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

flo-wer committed:
always export max_age/min_age

flo-wer committed:
downgrade errors of missing data
on start with empty recorder database these errors are logged multiple times:
ERROR (MainThread) [homeassistant.components.sensor.statistics] mean requires at least one data point
ERROR (MainThread) [homeassistant.components.sensor.statistics] variance requires at least two data points

downgrade them to debug as they are not meaningful to end users

flo-wer committed:
add change_rate attribute
this calculates the average change rate of all data points

flo-wer committed:
simplify count, reorder attribute calculation

flo-wer committed:
reorder initialization

flo-wer committed:
reorder attribute names

flo-wer committed:
don't use min/max for min_age/max_age

flo-wer committed:
add test case

flo-wer commented:
If these changes are fine, I am also going to write documentation for all attributes.

I also wanted to remove the `round(..., 2) ` of mean, median, stdev, variance and total as the hard coded length of 2 decimal places can be bad if the values of the original sensor are small. Is this acceptable?

flo-wer committed:
style

flo-wer committed:
style

fabaff commented on homeassistant/components/sensor/statistics.py:
```diff
@@ -84,13 +85,13 @@ def __init__(self, hass, entity_id, name, sampling_size, max_age):
-        self.average_change = self.change = 0
-        self.max_age = self.min_age = 0
+        self.count = 0
+        self.mean = self.median = self.stdev = self.variance = STATE_UNKNOWN
```
Should be initialized as `None` and not `STATE_UNKNOWN`. Home Assistant is handling it.

flo-wer replied:
done

fabaff commented on homeassistant/components/sensor/statistics.py:
```diff
@@ -183,36 +178,45 @@ def async_update(self):
+
+                    time_diff = (self.max_age - self.min_age).total_seconds()
+                    if time_diff > 0:
+                        self.change_rate = self.average_change / time_diff
```
This should be rounded.

flo-wer replied:
I think no values should be rounded at all as the decimal places depend on the original sensor value. The hard coded value 2 has bene chosen arbitrarily and for my precisely use case change_rate has at least 2 zero decimal places which would result in an always zero value.

These are 2 examples of my sensor that calculates the change rate of a water level based on a level of 100%:
```
average_change: -0.12000000000000455
change: -0.2400000000000091
change_rate: -0.004051726227830205
count: 3
max_age: 2018-08-10T03:50:05.168733+00:00
max_value: 99.84
min_age: 2018-08-10T03:49:35.551727+00:00
min_value: 99.6
sampling_size: 3
total: 299.2
unit_of_measurement: %

average_change: 0.11999999999999744
change: 0.23999999999999488
change_rate: 0.0013346913966593103
count: 3
max_age: 2018-08-10T03:52:15.271830+00:00
max_value: 99.44
min_age: 2018-08-10T03:50:45.363406+00:00
min_value: 99.2
sampling_size: 3
total: 298
unit_of_measurement: %
```
change_rate is then used to calculate a rate in L/min which results in a large enough value (above example ~0.9 L/min) but this would be to specific for a general sensor, so I chose seconds instead of minutes.

I don't use the other attributes but I imagine this could be the case for others too. So I would propose to remove the round for all values and just export the raw values.  Also most times you still have to use a template sensor to access the actual value, so everyone can round to his specific needs.

fabaff replied:
If you are exposing a value to the frontend then the amount of decimal places must be limited. 10 and more decimal places will not make the result more accurate or better readable. We don't care about the amount of decimal places used internally but for the frontend.

You will always find uses cases where the amount of decimal places or the unit of measurement doesn't match your needs. For all those cases we have the `template` sensor and `custom_components` which let you do exactly what you want.

fabaff commented on homeassistant/components/sensor/statistics.py:
```diff
@@ -183,36 +178,45 @@ def async_update(self):
-                if self._max_age is not None:
-                    self.max_age = max(self.ages)
-                    self.min_age = min(self.ages)
+
```
This should be rounded.

fabaff commented on homeassistant/components/sensor/statistics.py:
```diff
@@ -143,26 +142,22 @@ def should_poll(self):
+                ATTR_MIN_VALUE: self.min,
+                ATTR_MAX_VALUE: self.max,
+                ATTR_MIN_AGE: self.min_age,
+                ATTR_MAX_AGE: self.max_age,
```
If the configuration option is not set to use `max_age` then `max_age`/`min_age` should not be exposed. The timestamps are not useful because it's the timestamp for the first value and the last sensor update will write a new timestamp. And this timestamp is the same as the one stored in the entity state.

flo-wer replied:
To display the timestamp of the oldest/newest sensor value is the single use case for this, as this information is not available in e.g. a template sensor to calculate the time range, etc.

My first intention only was to always export the `max_age`/`min_age` and calculate change_rate in a template sensor but I then added this as a feature directly. I see no difference in always exporting these values and only exporting them when `max_age` is specified. They always have the same meaning. On the other hand, exporting `max_age`/`min_age` when `max_age` is used seems the less useful case as `max_age` (the config option) implies a approximation for `max_age` (and `min_age` for `now()`) where using only sampling_size doesn't give any approximation or even a rough scale for any timestamp.

flo-wer committed:
sort constants

flo-wer committed:
init variables with None

flo-wer commented on homeassistant/components/sensor/statistics.py:
```diff
@@ -23,19 +23,20 @@
+ATTR_MAX_AGE = 'max_age'
+ATTR_CHANGE = 'change'
+ATTR_AVERAGE_CHANGE = 'average_change'
+ATTR_CHANGE_RATE = 'change_rate'
```
done

flo-wer committed:
add precision config setting

flo-wer committed:
round to precision

flo-wer committed:
test round

flo-wer commented:
I have now added a new `precision` setting following the Filter Sensor. It can be used to specify how many decimal places are exported. I also added rounding to all remaining values.

flo-wer commented:
Is there something missing to finally get this merged? That is waiting for two weeks now without any progress.
